### PR TITLE
feat: add chainsaw based e2e tests

### PR DESCRIPTION
**Pull request by eddycharly**
## What

This PR adds chainsaw based e2e tests (https://github.com/kyverno/chainsaw)

## How

Added a couple of chainsaw tests and updated ci to run those tests.

## Breaking Changes

NONE

### CODE REVIEW

**Comment by eddycharly** [comment]
Talked about that with someone at KubeCon today, let's discuss further tomorrow =)

**Comment by jonstacks** [comment]
@eddycharly, thank you for stopping by the booth. I still need to talk with the rest of our team, but I think we will end up taking this and gradually build up more chainsaw tests.

Just linking the product page for others on our team as I found it helpful: https://kyverno.github.io/chainsaw/latest/

**Comment by eddycharly** [comment]
Thanks @jonstacks take your time and i hope you will find chainsaw useful.
BTW the nix package is [here](https://search.nixos.org/packages?channel=unstable&from=0&size=50&sort=relevance&type=packages&query=kyverno-chainsaw).

**Comment by hjkatz** [comment]
Wow thank you @eddycharly for this addition! I'll be looking to try and use these tests today and see if we can get this merged.

e2e tests like these will really help us reach stability 😄

File: Makefile [hunk 1]
```diff
@@ -277,3 +277,9 @@ helm-update-snapshots: _helm_setup ## Update helm unittest snapshots
 
 helm-update-snapshots-no-deps: ## Update helm unittest snapshots without rebuilding dependencies
 	$(MAKE) -C $(HELM_CHART_DIR) update-snapshots
+
+##@ Chainsaw
+
+.PHONY: chainsaw-tests
+chainsaw-tests: ## Run chainsaw tests
+	chainsaw test ./tests/chainsaw
```
Comment (hjkatz) [inline Makefile:285]: <i class='graphite__hidden'>[Re: lines +281 to +286]<br/><br/></i>

Nit: imo Chainsaw is the implementation of end-to-end tests, so perhaps we call this target `e2e-tests` in-case we change the implementation later?<span class='graphite__hidden'><br/><br/>See this comment inline on <a href="https://app.graphite.dev/github/pr/ngrok/ngrok-operator/506?utm_source=unchanged-line-comment">Graphite</a>.</span>

File: .github/workflows/ci.yaml [hunk 2]
```diff
@@ -153,6 +153,17 @@ jobs:
         kubectl get nodes
         kubectl get pods -A
 
+    - name: Install cosign
```
Comment (hjkatz) [inline .github/workflows/ci.yaml:156]: <i class='graphite__hidden'>[Re: lines +148 to +156]<br/><br/></i>

Q. I think I'm reading this as our `make deploy` will control the deployment config for the chainsaw tests step just below. Does this mean we need different sets of workflows for different sets of chainsaw tests? (for example enabling gateway or bindings features, or changing the behaviour of the ingress features)

Perhaps we could use different `try -> apply -> file` to apply different sets of manifests generated by `helm template`, then cleanup the operator in-between tests?<span class='graphite__hidden'><br/><br/>See this comment inline on <a href="https://app.graphite.dev/github/pr/ngrok/ngrok-operator/506?utm_source=unchanged-line-comment">Graphite</a>.</span>

**Comment by hjkatz** [comment]
@eddycharly I would like to play around with this locally and submit some changes for nix. Though I'm not exactly sure how to collaborate on this branch... perhaps I create my own fork and submit PRs to your fork so we can update this PR for our main fork? 😂 seems like a lot but maybe that would work?

File: Makefile (hunk 1 above)
Comment (eddycharly) [reply Makefile]: Done

File: .github/workflows/ci.yaml (hunk 2 above)
Comment (eddycharly) [reply .github/workflows/ci.yaml:156]: There's different solutions for this, in kyverno we use a GH job matrix to deploy kyverno with different configs.

We can meet today if you want and look at that together.

**Comment by eddycharly** [comment]
> Would you kindly enable this setting so we could collaborate?

@hjkatz i'll come to the booth with my laptop, will be easier ;)

**Comment by hjkatz** [comment]
I'm unfortunately at kubecon - But I'm sure @jonstacks will love to help out.

FWIW I think you did have this setting enabled because I was able to push a commit to your fork's branch: https://github.com/ngrok/ngrok-operator/pull/506/commits (2nd commit at the bottom)

**Review by Alice-Lilith — COMMENTED** [review]
Huh... yaml driven k8s tests. This is neat 👀

**Comment by eddycharly** [comment]
Hi folks, i'm back at work. Let me know if you consider merging this PR and use chainsaw, i can help with changes if you want.
cc @jonstacks @hjkatz @Alice-Lilith

**Comment by hjkatz** [comment]
@eddycharly Sorry for the delay - We're are excited to merge this and I'll be focused on it today.

Q. Do you have any additions for the PR before we merge?

File: .github/workflows/ci.yaml (hunk 2 above)
Comment (hjkatz) [reply .github/workflows/ci.yaml:156]: Talked with @jonstacks and we'll do a deploy of all features enabled so we can generally test a swath of the expected behaviour for now. Then we can come back with more specific cases if desired.

**Review by hjkatz — APPROVED** [review]

**Comment by eddycharly** [comment]
> Do you have any additions for the PR before we merge?

Not really, next step would be to add more tests, maybe i will check the issues to see if they can be converted to a test easily.
In Kyverno user reporting issues usually contain manifests to reproduce the issue and it's often very easy to turn those into a test, maybe it's the same here 🤞

**Comment by eddycharly** [comment]
BTW we have a dedicated slack channel you can use if you have questions https://kubernetes.slack.com/archives/C067LUFL43U

**Comment by eddycharly** [comment]
@hjkatz either the pod doesn't start or keeps restarting because of invalid api-key/token 🤔 
I can look at it later if it helps, you can also temporarily disable this test until you come up with a solution for the api-key/token.

**Comment by hjkatz** [comment]
hah, sorry for any emails - We're iterating on this right now trying to figure out exactly how we want these to run.

**Comment by hjkatz** [comment]
@eddycharly Sorry for all of the crazy noise! This is finally ready to merge. There were some adjustments @jonstacks and I decided to make so the e2e tests will work for both ngrok-op/main and forks. tl; dr:
* Separate out changes and build-and-test into composite github actions
* Create 2 workflows: `ci.yaml` (pull_request) and `test.yaml` (pull_request_target)
* Ensure that each workflow has access to secrets (or not) as needed

**Comment by eddycharly** [comment]
@hjkatz awesome !

I hope you like chainsaw, don't hesitate to ask questions on our slack channel if needed or file issues or contribute to the project if you want.

Additionally you can add yourself in the [ADOPTERS](https://github.com/kyverno/chainsaw/blob/main/ADOPTERS.md) file if you want.

I didn't have much time to help but i will try to search the issues and see what tests it makes sense to add when i have more time.

**Pull request merged**